analysers/analyser_osmosis_powerline.py

```diff
@@ -211,57 +298,112 @@
+# It looks for voltage continuation on every junction. Two (or more) topoedges on a given node with the same voltage means a connection.
+# TODO support partial termination (i.e termination|straight) with different voltages involved.
 sql30 = """
-CREATE VIEW power_lines_junctions AS
-SELECT
-    p.nid
-FROM
-    (SELECT nid FROM power_lines_nodes n WHERE n.voltage IS NOT NULL GROUP BY n.wid, n.nid) AS p
-GROUP BY
-    p.nid
-HAVING
-    COUNT(*) > 1
-"""
+WITH nodes_voltage AS (
+    SELECT
+        nid,
+        tid,
+        unnest(voltage)::varchar AS voltage
+    FROM
+        power_lines_topoedges
+),
+nodes_voltage_values AS (
+    SELECT
+        nid,
+        tid,
+        voltage,
+        round((voltage::numeric / 1000)::numeric,1)::varchar AS voltage_val,
+        'numeric' AS origin
+    FROM
+        nodes_voltage
+    WHERE
+        voltage ~ '^[0-9.]+$'
+
+    UNION
+
+    SELECT
+        nid,
+        tid,
+        voltage AS voltage,
+        round((voltage::numeric / (1000 * sqrt(3)))::numeric,1)::varchar AS voltage_val,
+        'numeric' AS origin
+    FROM
+        nodes_voltage
+    WHERE
+        voltage ~ '^[0-9.]+$'
+
+    UNION
+
+    SELECT
+        nid,
+        tid,
+        voltage AS voltage,
+        voltage AS voltage_val,
+        'varchar' AS origin
+    FROM
+        nodes_voltage
+    WHERE
+        NOT(voltage ~ '^[0-9.]+$')
+),
+
+nodes_selected AS (
+    SELECT
+        nid
+    FROM
+        power_lines_topoedges
+    GROUP BY
+        nid
+    HAVING
+        COUNT(distinct tid) > 1
+),
+voltage_groups AS (
+    SELECT
+        n.nid,
+        max(n.voltage) AS voltage,
+        n.voltage_val,
+        count(n.voltage) AS cv,
+        n.origin
+    FROM
+        nodes_voltage_values AS n
+        JOIN nodes_selected AS s ON
+            s.nid = n.nid
+    GROUP BY
+        n.nid, n.voltage_val, n.origin
+)
 
-# Every junctions that aren't transformers cross, splits or terminations repeated a single time (meaning the junction involves different voltages)
-sql31 = """
 SELECT
-    DISTINCT(j.nid),
+    DISTINCT(v.nid),
     ST_AsText(nodes.geom)
 FROM
-    power_lines_junctions j
-    NATURAL JOIN power_lines_nodes n
+    voltage_groups AS v
     JOIN nodes ON
-        n.nid = nodes.id
+        v.nid = nodes.id
 WHERE
-    n.voltage is not null AND
     (
         NOT nodes.tags?'power' OR
         nodes.tags->'power' != 'transformer'
     ) AND
     NOT nodes.tags?'transformer' AND -- example: power=pole + transformer=*
     (
-        NOT nodes.tags?'line_management' OR
-        (
-            NOT 'split' = ANY(string_to_array(nodes.tags->'line_management', '|')) AND
-            NOT 'termination' = ANY(string_to_array(nodes.tags->'line_management', '|')) AND
-            nodes.tags->'line_management' != 'cross'
-        )
+        NOT nodes.tags?'line_management' OR (nodes.tags->'line_management' != 'cross' AND nodes.tags->'line_management' != 'termination')
```

Comment:
Please, keep the AND and OR formatting, all the Osmose code is done as is.

Reply:
It's done